### PR TITLE
Label encoder opset4

### PR DESCRIPTION
**Pull request by adityagoel4512**
### Description

Implements LabelEncoder as per `ai.onnx.ml` opset 4 for the upcoming ONNX 1.15 release. ~~This currently depends on a new ONNX release candidate and so is marked as draft in the meantime.~~

### Motivation and Context
Closes https://github.com/microsoft/onnxruntime/issues/17602

### CODE REVIEW

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 1]
```diff
@@ -105,5 +106,80 @@ class LabelEncoder_2 final : public OpKernel {
...
+    }
+
+    return Status::OK();
+  }
+
+ private:
+  void InitializeAttrFields(const OpKernelInfo& kernel_info);
+  InlinedHashMapNaNSensitive<TKey, TValue> _map;
+  TValue _default_value;
+  std::string _key_field_name;
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:180]: Names with leading underscored are reserved. Please, use trailing underscore to designate a data member.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 2]
```diff
@@ -105,5 +106,80 @@ class LabelEncoder_2 final : public OpKernel {
...
 };
+
+template <typename T>
+std::vector<T> GetAttribute(const OpKernelInfo& info, const std::string& name, const std::string& tensor_name) {
+  std::vector<T> attrs;
+  auto result = info.GetAttrs<T>(name, attrs);
+  if (!result.IsOK()) {
+    ONNX_NAMESPACE::TensorProto attr_tensor_proto;
+    result = info.GetAttr(tensor_name, &attr_tensor_proto);
+    ORT_ENFORCE(result.IsOK(), "LabelEncoder is missing an attribute");
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:117]: The error message must include details about the name and the type of the attribute.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 3]
```diff
@@ -105,5 +106,80 @@ class LabelEncoder_2 final : public OpKernel {
...
+  }
+  return attrs;
+}
+
+template <typename T>
+T GetDefault(const OpKernelInfo& info, const std::string& attr_name, const T& backup) {
+  ONNX_NAMESPACE::TensorProto attr_tensor_proto;
+  auto result = info.GetAttr("default_tensor", &attr_tensor_proto);
+  if (result.IsOK() && utils::HasDataType(attr_tensor_proto)) {
+    auto default_value = utils::ParseData<T>(attr_tensor_proto);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:128]: Use Unpack data

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 4]
```diff
@@ -105,5 +106,80 @@ class LabelEncoder_2 final : public OpKernel {
...
+    if (nullptr == X) {
+      return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
+    }
+    const TensorShape& shape = X->Shape();
+    auto* Y = context->Output(0, shape);
+
+    auto input = X->template DataAsSpan<TKey>();
+    auto output = Y->template MutableDataAsSpan<TValue>();
+
+    for (int64_t i = 0; i < shape.Size(); ++i) {
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h:165]: Another suggestion is not to use index, but simply iterate input and output pointers (span iterators).

**Review by yuslepukhin — CHANGES_REQUESTED** [review]

File: include/onnxruntime/core/common/inlined_containers.h [hunk 5]
```diff
@@ -75,6 +75,56 @@ class InlinedHashMap : public absl::flat_hash_map<Key, Value,
...
+
+template <typename T>
+struct NaNEqual {
+  bool operator()(const T& lhs, const T& rhs) const {
+    return absl::container_internal::hash_default_eq<T>{}(lhs, rhs);
+  }
+};
+
+template <>
+struct NaNEqual<float> {
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply include/onnxruntime/core/common/inlined_containers.h:104]: maybe add additional template parameters to InlinedHashMap with defaults that get passed through to absl::flat_hash_map? then InlinedHashMap can be further customized elsewhere.

[the remark this answers is not in the document]
Comment (adityagoel4512) [reply include/onnxruntime/core/common/inlined_containers.h]: Did the latter approach and added template parameters.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h]: This is actually how the existing LabelEncoder compute method was written. I updated this as well to iterate the pointers instead.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h (hunk 1 above)
Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h]: Again, this was the existing code style in the file. I've updated the existing code to conform too.

**Comment by thiagocrepaldi** [comment]
@adityagoel4512 any update on this? ORT 1.17 release is coming soon :)

File: include/onnxruntime/core/common/inlined_containers_fwd.h [hunk 6]
```diff
@@ -23,6 +23,7 @@
 #endif  // _MSC_VER
 
 #include <absl/container/inlined_vector.h>
+#include <absl/container/flat_hash_map.h>
```
Comment (yuslepukhin) [inline include/onnxruntime/core/common/inlined_containers_fwd.h:26]: The problem with this, at least in the past, not everything compiles under CUDA. That is why a FWD header was invented to be used in headers as a forward declaration.

FWD headers minimize inclusions. I understand that this is needed for default arguments, but it leads to huge problems down the road.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 7]
```diff
@@ -53,57 +54,169 @@ class LabelEncoder_2 final : public OpKernel {
...
-    ORT_ENFORCE(num_keys == num_values,
-                "The ", _key_field_name, " and ", _value_field_name, " attribtues in LabelEncoder ",
-                "(name: ", info.node().Name(), ") must have the same length. ",
-                "However, the number of key is ", num_keys, " and the number of ",
-                "values is ", num_values, ".");
-    _map.reserve(num_keys);
-    for (size_t i = 0; i < num_keys; ++i)
-      _map.emplace(keys[i], values[i]);
+    ORT_ENFORCE(num_keys == num_values, "The ", key_field_name_, " and ", value_field_name_,
+                " attribtues in LabelEncoder ", "(name: ", info.node().Name(), ") must have the same length. ",
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:63]: >attribtues [](http://example.com/codeflow?start=18&length=10)

Typo

File: onnxruntime/core/providers/cpu/ml/label_encoder.h (hunk 1 above)
Comment (yuslepukhin) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h]: We do not want to multiply code that is not conformant.

File: onnxruntime/core/providers/cpu/ml/label_encoder.cc [hunk 8]
```diff
@@ -55,168 +52,267 @@ Status LabelEncoder::Compute(OpKernelContext* context) const {
...
     LabelEncoder_2<std::int64_t, std::int64_t>)
 
 template <>
 void LabelEncoder_2<std::int64_t, std::int64_t>::InitializeSomeFields(const OpKernelInfo& info) {
-  _key_field_name = "keys_int64s";
-  _value_field_name = "values_int64s";
-  info.GetAttrOrDefault<std::int64_t>("default_int64", &_default_value, (std::int64_t)-1);
+  key_field_name_ = "keys_int64s";
+  value_field_name_ = "values_int64s";
+  info.GetAttrOrDefault<std::int64_t>("default_int64", &default_value_, (std::int64_t)-1);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.cc:188]: >std::int64_t)- [](http://example.com/codeflow?start=73&length=14)

No C casts

File: onnxruntime/core/providers/cpu/ml/label_encoder.cc [hunk 9]
```diff
@@ -55,168 +52,267 @@ Status LabelEncoder::Compute(OpKernelContext* context) const {
...
+                                                      std::vector<MLDataType>{DataTypeImpl::GetTensorType<float>()})
+                                      .TypeConstraint("T2",
+                                                      std::vector<MLDataType>{DataTypeImpl::GetTensorType<float>()}),
+                                  LabelEncoder_4<float, float>)
+
+template <>
+void LabelEncoder_4<float, float>::InitializeAttrFields(const OpKernelInfo& kernel_info) {
+  key_field_name_ = "keys_floats";
+  value_field_name_ = "values_floats";
+  default_value_ = GetDefault(kernel_info, "default_float", static_cast<float>(-0.f));
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.cc:245]: redundant cast

File: onnxruntime/core/providers/cpu/ml/label_encoder.cc [hunk 10]
```diff
@@ -55,168 +52,267 @@ Status LabelEncoder::Compute(OpKernelContext* context) const {
...
+    KernelDefBuilder()
+        .TypeConstraint("T1", std::vector<MLDataType>{DataTypeImpl::GetTensorType<std::int64_t>()})
+        .TypeConstraint("T2", std::vector<MLDataType>{DataTypeImpl::GetTensorType<float>()}),
+    LabelEncoder_4<std::int64_t, float>)
+
+template <>
+void LabelEncoder_4<std::int64_t, float>::InitializeAttrFields(const OpKernelInfo& kernel_info) {
+  key_field_name_ = "keys_int64s";
+  value_field_name_ = "values_floats";
+  default_value_ = GetDefault(kernel_info, "default_float", static_cast<float>(-0.f));
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.cc:230]: >static_cast<float> [](http://example.com/codeflow?start=60&length=18)

Redundant cast

File: onnxruntime/core/providers/cpu/ml/label_encoder.cc (hunk 9 above)
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.cc:245]: >static_cast< [](http://example.com/codeflow?start=60&length=12)

Ditto

File: onnxruntime/core/providers/cpu/ml/label_encoder.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h]: These are actual bugs. Need fixing.

File: onnxruntime/core/providers/cpu/ml/label_encoder.cc (hunk 8 above)
Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/ml/label_encoder.cc:188]: Again, this was existing code but good point.

File: include/onnxruntime/core/common/inlined_containers_fwd.h (hunk 6 above)
Comment (gramalingam) [reply include/onnxruntime/core/common/inlined_containers_fwd.h:26]: Is there a specific alternative you recommend? (Sorry, I don't know the context enough to understand, maybe this is a naive question.)

Comment (yuslepukhin) [reply include/onnxruntime/core/common/inlined_containers_fwd.h:26]: I have not had time to think about it for longer. Ed's suggestion looks good as far as the templates go. However, if the next version of absl header fails to compile some place, we are screwed.

**Comment by yuslepukhin** [comment]
/asp run Windows x64 QNN CI Pipeline, orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-ortmodule-distributed

File: include/onnxruntime/core/common/inlined_containers_fwd.h (hunk 6 above)
Comment (yuslepukhin) [reply include/onnxruntime/core/common/inlined_containers_fwd.h:26]: I can see CUDA build is failing due to ABSL code.

Comment (yuslepukhin) [reply include/onnxruntime/core/common/inlined_containers_fwd.h:26]: My specific suggestion is to use `std` container for the kernel and explain the kernel why and revert changes in the common container's headers. We would need to seek a long-term solution.

Comment (adityagoel4512) [reply include/onnxruntime/core/common/inlined_containers_fwd.h]: The abseil flat_hash_map uses noticeably less memory with large LabelEncoders so I would like to retain it where possible. I opted to revert the container header changes and simply localise the definition of the hashmap in "labelencoder.h" where we already had an `#ifndef DISABLE_ABSEIL`.

**Comment by adityagoel4512** [comment]
Note, I added a string -> int16 mapping because the ONNX backend tests [require](https://github.com/onnx/onnx/blob/main/onnx/backend/test/case/node/ai_onnx_ml/label_encoder.py#L51-L100) it.

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 11]
```diff
@@ -53,57 +54,181 @@ class LabelEncoder_2 final : public OpKernel {
...
+
+    auto input = X->template DataAsSpan<TKey>();
+    auto output = Y->template MutableDataAsSpan<TValue>();
+    auto input_iter = input.begin();
+    auto output_iter = output.begin();
+    while (input_iter != input.end()) {
+      const auto found = map_.find(*input_iter);
+      *output_iter = found == map_.end() ? default_value_ : found->second;
+      output_iter++;
+      input_iter++;
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:82]: Last comment, let's make it pre-increment. Takes place for every tensor element may subtly affect perf, which may even not show on a radar.

**Comment by yuslepukhin** [comment]
There is a compile error that affects multiple pipelines.

**Review by yuslepukhin — DISMISSED** [review]
:shipit:

**Comment by yuslepukhin** [comment]
Lint is needed

> 2024-01-12T00:12:04.7661594Z          C:\a\_work\1\s\onnxruntime\core/providers/cpu/ml/label_encoder.h(121,20): warning C4244: '*=': conversion from 'const Element' to 'size_t', possible loss of data [C:\a\_work\1\b\RelWithDebInfo\onnxruntime_providers.vcxproj]

File: onnxruntime/core/providers/cpu/ml/label_encoder.h [hunk 12]
```diff
@@ -53,57 +54,181 @@ class LabelEncoder_2 final : public OpKernel {
...
+  ONNX_NAMESPACE::TensorProto attr_tensor_proto;
+  auto result = info.GetAttr(tensor_name, &attr_tensor_proto);
+  if (name.empty()) {
+    ORT_ENFORCE(result.IsOK(), "LabelEncoder is missing attribute ", tensor_name);
+  } else {
+    ORT_ENFORCE(result.IsOK(), "LabelEncoder is missing attribute ", tensor_name, " or ", name);
+  }
+  size_t tensor_size = 1;
+  for (auto dim : attr_tensor_proto.dims()) {
+    tensor_size *= static_cast<size_t>(dim);
```
Comment (yuslepukhin) [inline onnxruntime/core/providers/cpu/ml/label_encoder.h:121]: >tensor_size *= static_cast<size_t>(dim); [](http://example.com/codeflow?start=4&length=40)

We need to compute the result first in int64_t,
then use `narrow<size_t>()` to make sure any data loss is detected.

Comment (adityagoel4512) [reply onnxruntime/core/providers/cpu/ml/label_encoder.h]: Thanks for the suggestion!

**Review by yuslepukhin — DISMISSED** [review]

**Comment by yuslepukhin** [comment]
Lint is failiing.
In your enlistment install lintrunner
lintrunner init
lintrunner -a

**Comment by adityagoel4512** [comment]
I already have been - to not pollute the diff I assumed the CI only considered the diff for linting but apparently not.

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Comment by yuslepukhin** [comment]
Please, look at the following issues of lint.
https://github.com/microsoft/onnxruntime/actions/runs/7497896828/job/20435289014?pr=17977

**Comment by YUNQIUGUO** [comment]
removing the tag as the commit has already been included in the rel-1.17.0 branch

**Pull request merged**